ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1244 @@
+/* SSL_get_wpoll_descriptor */
+int ossl_quic_get_wpoll_descriptor(QUIC_CONNECTION *qc, BIO_POLL_DESCRIPTOR *desc)
 {
-    return -1;
+    if (desc == NULL || qc->net_wbio == NULL)
+        return 0;
+
+    return BIO_get_wpoll_descriptor(qc->net_wbio, desc);
 }
 
-int ossl_quic_write(SSL *s, const void *buf, size_t len, size_t *written)
+/* SSL_want_net_read */
+int ossl_quic_get_want_net_read(QUIC_CONNECTION *qc)
 {
-    BIO *wbio = SSL_get_wbio(s);
-    int ret;
-    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_QUIC_SSL(s);
+    if (qc->ch == NULL)
+        return 0;
 
-    if (sc == NULL || wbio == NULL)
+    return ossl_quic_reactor_want_net_read(ossl_quic_channel_get_reactor(qc->ch));
+}
+
+/* SSL_want_net_write */
+int ossl_quic_get_want_net_write(QUIC_CONNECTION *qc)
+{
+    if (qc->ch == NULL)
         return 0;
 
-    sc->rwstate = SSL_WRITING;
-    ret = BIO_write_ex(wbio, buf, len, written);
-    if (ret > 0 || !BIO_should_retry(wbio))
-        sc->rwstate = SSL_NOTHING;
-    return ret;
+    return ossl_quic_reactor_want_net_write(ossl_quic_channel_get_reactor(qc->ch));
 }
 
+/*
+ * QUIC Front-End I/O API: Connection Lifecycle Operations
+ * =======================================================
+ *
+ *         SSL_do_handshake         => ossl_quic_do_handshake
+ *         SSL_set_connect_state    => ossl_quic_set_connect_state
+ *         SSL_set_accept_state     => ossl_quic_set_accept_state
+ *         SSL_shutdown             => ossl_quic_shutdown
+ *         SSL_ctrl                 => ossl_quic_ctrl
+ *   (BIO/)SSL_connect              => ossl_quic_connect
+ *   (BIO/)SSL_accept               => ossl_quic_accept
+ *
+ */
+
+/* SSL_shutdown */
 int ossl_quic_shutdown(SSL *s)
 {
+    QUIC_CONNECTION *qc = QUIC_CONNECTION_FROM_SSL(s);
+
+    if (!expect_quic_conn(qc))
+        return 0;
+
+    if (qc->ch != NULL)
+        ossl_quic_channel_local_close(qc->ch);
+
     return 1;
 }
 
+/* SSL_ctrl */
+static void fixup_mode_change(QUIC_CONNECTION *qc)
+{
+    /* If enabling EPW mode, cancel any AON write */
+    if ((qc->ssl_mode & SSL_MODE_ENABLE_PARTIAL_WRITE) != 0)
+        aon_write_finish(qc);
+}
+
 long ossl_quic_ctrl(SSL *s, int cmd, long larg, void *parg)
 {
-    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_QUIC_SSL(s);
+    QUIC_CONNECTION *qc = QUIC_CONNECTION_FROM_SSL(s);
 
-    if (sc == NULL)
+    if (!expect_quic_conn(qc))
         return 0;
 
-    switch(cmd) {
-    case SSL_CTRL_CHAIN:
-        if (larg)
-            return ssl_cert_set1_chain(sc, NULL, (STACK_OF(X509) *)parg);
-        else
-            return ssl_cert_set0_chain(sc, NULL, (STACK_OF(X509) *)parg);
+    switch (cmd) {
+    case SSL_CTRL_MODE:
+        qc->ssl_mode |= (uint32_t)larg;
+        fixup_mode_change(qc);
+        return qc->ssl_mode;
+    case SSL_CTRL_CLEAR_MODE:
+        qc->ssl_mode &= ~(uint32_t)larg;
+        fixup_mode_change(qc);
+        return qc->ssl_mode;
+    default:
```

Comment:
Shouldn't we forward unrecognised ctrls to the SSL_CONNECTION? Or possibly that's something I need to do in my PR...

Reply:
Agreed. Though I suspect it would make sense for you to do this in your PR.

Reply:
There are also some SSL_MODEs that need to be forwarded.